### PR TITLE
Generate test loom data with Hypothesis

### PR DESCRIPTION
**Pull request by MaybeJustJames**
Fix related bugs in API and especially implementation of
search_space::add_markers where element_type == "marker_gene": do not
assume that clustering ids are indexes.

### CODE REVIEW

File: server/scopeserver/dataserver/utils/search_space.py [hunk 1]
```diff
@@ -154,14 +154,9 @@ def add_markers(self, element_type="regulon_target") -> None:
...
-                x.split("_")[-1] for x in loom.ra.keys() if bool(re.search("ClusterMarkers_[0-9]+$", x))
-            ]
-            for clustering in searchable_clustering_ids:
-                clustering = int(clustering)
-                for cluster in range(len(self.meta_data["clusterings"][clustering]["clusters"])):
-                    genes = self.loom.get_cluster_marker_genes(clustering, cluster)
-                    for gene in genes:
+            for clustering in self.meta_data["clusterings"]:
+                for cluster in clustering["clusters"]:
+                    for gene in self.loom.get_cluster_marker_genes(clustering["id"], cluster["id"]):
```
Comment (KrisDavie) [inline server/scopeserver/dataserver/utils/search_space.py:159]: Any changes in the search space should also increment `CURRENT_SS_VERISON` on line 12

File: server/test/test_labels.py [hunk 2]
```diff
@@ -1,73 +1,67 @@
...
+def test_label_all_clusters(data):
+    "Test that labels are generated for each cluster in a clustering."
+    with loom_generator(data) as loom:
+        for clustering_meta in data.clusterings.metadata["clusterings"]:
+            cluster_names = {meta["id"]: meta["description"] for meta in clustering_meta["clusters"]}
+            labels = []
+            for cluster_id, cluster in cluster_names.items():
+                coords = loom.get_coordinates(-1, cluster_info=(clustering_meta["id"], cluster_id))
+                cluster_coords = Coordinate(np.mean(coords["x"]), y=np.mean(coords["y"]))
+                colour = constant.BIG_COLOR_LIST[cluster_id]
```
Comment (KrisDavie) [inline server/test/test_labels.py:59]: This should loop the list for when `cluster_id > len(constant.BIG_COLOR_LIST)`

**Review by KrisDavie — CHANGES_REQUESTED** [review]
Really nice work, only real things I'd need fixing are the search space version and the issue with a large cluster_id and the len of BIG_COLOR_LIST

File: server/scopeserver/dataserver/utils/search_space.py (hunk 1 above)
Comment (MaybeJustJames) [reply server/scopeserver/dataserver/utils/search_space.py:159]: Why would this change warrant a `CURRENT_SS_VERSION` bump?

File: server/test/generate.py [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+import hypothesis.extra.numpy
+
+import numpy as np
+import loompy
+
+
+from scopeserver.dataserver.utils.loom import Loom
+from scopeserver.dataserver.utils.loom_file_handler import LoomFileHandler
+
+LARGEST: int = 64
```
[the remark this answers is not in the document]
Comment (MaybeJustJames) [reply server/test/generate.py:25]: Would you find, `MAX_SIZE` better?

[the remark this answers is not in the document]
Comment (MaybeJustJames) [reply server/test/generate.py:25]: Or `SIZE_LIMIT`?

**Review by KrisDavie — APPROVED** [review]
Great!

**Pull request merged**